Aegis TUI integration

DS-KoolAid opened the pull request:
## Summary (required)

Adding Aegis interactive to the cli to improve the QOL for operators.

## JIRA (required)

- CHA-2926 and CHA-2927


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Interactive Aegis TUI plus CLI commands to list agents, view details, select agents, SSH into agents, browse/run jobs and capabilities, with contextual help and autocompletion.
* **Tests**
  * Added comprehensive TUI/unit tests covering list, set, info, ssh, and job flows.
* **Documentation**
  * Help screens, examples, and module docstrings for the Aegis UI/commands.
* **Chores**
  * Bumped version to 2.2.1, added textual/rich/prompt_toolkit dependencies, and enabled running the new TUI test suite.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

michaelweber commented:
Hold off on merging, doing one more round of testing.

michaelweber commented:
Closing for now since this is a nice-to-have versus a requirement.

DS-KoolAid committed:
adding aegis tui

MarioBartolome committed:
Bugfix: Fixed crash on health_check returning emtpy (None)

MarioBartolome committed:
Bugfix: Agent status indicator fix

MarioBartolome committed:
Refactored how the agents list is generated

MarioBartolome committed:
Refactored logic and styles

MarioBartolome committed:
Command compl<TAB!!!>

DS-KoolAid committed:
adding better ui and job command

EvanLeleux commented on praetorian_cli/handlers/aegis.py:
```diff
@@ -0,0 +1,142 @@
+        agent = sdk.aegis.get_by_client_id(client_id)
+        if not agent:
+            click.echo(f"Error: No Aegis instance found with client ID: {client_id}", err=True)
+            sys.exit(1)
```
For each of these errors you can use `from praetorian_cli.handlers.utils import error` to maintain consistency with other handlers.

EvanLeleux commented on praetorian_cli/handlers/aegis.py:
```diff
@@ -0,0 +1,142 @@
+@click.option('-D', '--dynamic-forward', help='Dynamic port forwarding/SOCKS proxy (e.g., 1080)')
+@click.option('--key', '-i', help='SSH private key file')
+@click.option('--ssh-opts', help='Additional SSH options (e.g., "-o StrictHostKeyChecking=no")')
+def ssh(sdk, client_id, user, local_forward, remote_forward, dynamic_forward, key, ssh_opts):
```
Usually handlers are very very simple and we move all the logic to the sdk. So unless its intentionally not in the SDK it would be better to have this just call a single method in the `aegis.py` sdk entity. Then other users of the SDK can also use it in their scripts (if that is even possible?).

EvanLeleux commented on praetorian_cli/sdk/chariot.py:
```diff
@@ -235,6 +237,43 @@ def start_mcp_server(self, allowable_tools=None):
+                user_email = payload.get('email')
+            except Exception:
+                # If JWT decoding fails, fall back to the account parameter
+                user_email = self.keychain.account
```
Could you extract JWT functionality decoding functionality into another function. May be useful some other time and can just return a dict of the contents of it.

EvanLeleux replied:
Also would we want it to fallback to the account param? Does supplying the username actually do anything or does Okta authorize everything ultimately?

DS-KoolAid replied:
yea shouldnt fall back to account - okta is the main authenticator here

EvanLeleux commented on praetorian_cli/sdk/entities/aegis.py:
```diff
@@ -0,0 +1,46 @@
+        """ List Aegis agents """
+        try:
+            # Aegis agents use a different API endpoint
+            resp = self.api._make_request('GET', self.api.url('/aegis/agent'))
```
Use `Chariot.get` instead of directly using `_make_request`

EvanLeleux commented on praetorian_cli/sdk/entities/aegis.py:
```diff
@@ -0,0 +1,46 @@
+                    'network_interfaces': agent.get('network_interfaces', []),
+                    'health_check': agent.get('health_check')
+                }
+                formatted_agents.append(formatted_agent)
```
Could you just add the formatted `key` to the dict rather than recreating the entire dict

EvanLeleux commented on praetorian_cli/ui/aegis/commands/help_command.py:
```diff
@@ -0,0 +1,95 @@
+        """Execute help command"""
+        self.show_help()
+    
+    def show_help(self):
```
This seems long and hard to maintain. Could each subcommand implement a `show_help` method that this command just iterates through calling each subcommand's `show_help` method. That way a new command wouldnt have to edit this

EvanLeleux commented on praetorian_cli/ui/aegis/commands/info_command.py:
```diff
@@ -0,0 +1,201 @@
+        architecture = agent.get('architecture', 'Unknown')
+        fqdn = agent.get('fqdn', 'N/A')
+        client_id = agent.get('client_id', 'N/A')
+        last_seen = agent.get('last_seen_at', 0)
```
It seems this is repeated quite a lot over the whole PR, probably useful to centralize as an Agent type and function from_dict.

EvanLeleux replied:
Same with this snippet 
```
        cf_status = health['cloudflared_status']
        public_hostname = cf_status.get('hostname')
        authorized_users = cf_status.get('authorized_users', '')
        tunnel_name = cf_status.get('tunnel_name', 'N/A')
```

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+            self.show_latest_job()
+        else:
+            self.console.print(f"\n  Unknown job subcommand: {subcommand}")
+            self.show_job_help()
```
Does TUI provide groups to handle this or is it purely UI related. Just since it seems like this is pretty much what `click` does for our handlers

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+        """List jobs for the selected agent"""
+        if not self.selected_agent:
+            self.console.print("\n  No agent selected. Use 'set <id>' to select one.\n")
+            return
```
Could extract this function too or handle it centrally in the `execute` function.

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+            if capability.lower() not in valid_capabilities:
+                self.console.print(f"\n  [{self.colors['error']}]Invalid capability: '{capability}'[/{self.colors['error']}]")
+                self.console.print(f"  [{self.colors['dim']}]Use 'job capabilities' to see available options[/{self.colors['dim']}]\n")
+                return
```
Do we need to check this or does the API return an error if we try to run an unknown or invalid capability against an Aegis internal surface?

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+                        target_type = cap.get('target', 'asset').lower()
+                        break
+            except:
+                pass
```
You already queried the capabilities list above around line 160, could you reuse that here?

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+                self.console.print(f"  [{self.colors['dim']}]Target domain: {domain}[/{self.colors['dim']}]")
+            else:
+                target_key = f"#asset#{hostname}#{hostname}"
+                self.console.print(f"  [{self.colors['dim']}]Target asset: {hostname}[/{self.colors['dim']}]")
```
This doesnt seem too maintainable especially as we add more types that Aegis capabilities run off. Could we instead prompt the user to supply a key for now or in the future reuse the target type, query the target type, display the list of returned objects, and allow them to interactively view the list of objects to select. Similar to the capabilities page in the UI?

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+                    # Use capitalized field names to match UI format
+                    creds = {"Username": username, "Password": password}
+                    config.update(creds)
+                    self.menu._cached_creds[cache_key] = creds
```
We may be able to reuse the capabilities query from above and see if they have any required parameters instead of checking the cred_keywords. It also seems they may all have the same parameters `Username` and `Password`.

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+                    self.console.print(f"  ✗ No job returned from API")
+                    
+            except Exception as e:
+                self.console.print(f"  ✗ Error: {e}")
```
Probably dont need this try block since the exception would be caught by the try block above it.

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+            
+            if not all_capabilities:
+                self.console.print(f"  [{self.colors['dim']}]No Aegis capabilities found[/{self.colors['dim']}]\n")
+                return
```
Second time seeing this, probably cleaner if we move to a helper function `get_aegis_capabilities`

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+                assets_resp = self.sdk.assets.list(asset_type='addomain')
+                assets = assets_resp[0] if isinstance(assets_resp, tuple) else assets_resp
+                if isinstance(assets, dict):
+                    assets = assets.get('assets', assets.get('data', assets.get('items', [])))
```
This also seems to be common use, checking if API response is a tuple, extracting first, then extracting the dict. Probably could make a helper function `dict_from_resp` for this.

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+            self.console.print(f"     0. Enter domain manually")
+            
+            from rich.prompt import Prompt
+            max_choice = min(len(domains), 15)
```
Moving all import statements to the top of the file would reduce having to import things multiple times like in this case

EvanLeleux commented on praetorian_cli/ui/aegis/commands/job_command.py:
```diff
@@ -0,0 +1,660 @@
+            return Text("running", style=f"{self.colors['warning']}")
+        elif status_upper.startswith('JP'):
+            return Text("passed", style=f"{self.colors['success']}")
+        elif status_upper.startswith('JF'):
```
Extract these strings as globals in global.py

EvanLeleux commented on praetorian_cli/ui/aegis/commands/list_command.py:
```diff
@@ -0,0 +1,245 @@
+            # negative sorting (True first)
+            return (
+                0 if (is_online and has_tunnel) else (1 if is_online else 2),
+                -(last_seen_seconds or 0)
```
Breaking this out of a 1 liner may make it easier to read

EvanLeleux commented on praetorian_cli/ui/aegis/commands/ssh_command.py:
```diff
@@ -0,0 +1,270 @@
+        ssh_args = args or []
+        self.handle_ssh_command(ssh_args)
+    
+    def handle_ssh_command(self, ssh_args=None):
```
There seems to be a ton of code here to support the same functionality that `praetorian chariot aegis ssh` handles. Could you use a subprocess here and just call that?

EvanLeleux commented on praetorian_cli/ui/aegis/commands/tasks_command.py:
```diff
@@ -0,0 +1,26 @@
+from .base_command import BaseCommand
+
+
+class TasksCommand(BaseCommand):
```
If its pending we probably can just remove it entirely from the PR incase it changes before we actually implement it (task becomes command or something)

EvanLeleux commented on praetorian_cli/ui/aegis/menu.py:
```diff
@@ -0,0 +1,446 @@
+        
+        readline = DummyReadline()
+        if VERBOSE:
+            print("No readline available - tab completion disabled")
```
Could we just make `gnureadline` a requirement? It didnt seem to work for me otherwise

EvanLeleux replied:
It would also remove all the dummyreadline checks below

EvanLeleux commented on praetorian_cli/ui/aegis/menus/style.py:
```diff
@@ -0,0 +1,55 @@
+        """Get color by name"""
+        return self.colors.get(color_name, '#FFFFFF')
+    
+    def format_success(self, text: str) -> str:
```
It seems only the completion manager actually uses these helper functions while the commands do it themselves

EvanLeleux commented on praetorian_cli/ui/aegis/commands/help_command.py:
```diff
@@ -0,0 +1,95 @@
+            title="[bold]Help[/bold]",
+            title_align="left"
+        ))
+        self.pause()
```
Why do we pause after help?

EvanLeleux commented on praetorian_cli/ui/aegis/commands/list_command.py:
```diff
@@ -0,0 +1,245 @@
+        """Compose and display the agents table using pre-computed properties"""
+        if not self.agents:
+            no_agents_panel = Panel(
+                Align.center(f"[{self.colors['warning']}]No agents available[/{self.colors['warning']}]\n[dim]Press 'r' to reload[/dim]"),
```
I don't think pressing R reloads right?

EvanLeleux reviewed (changes requested):
My main concern is that this PR appears to reimplement `praetorian-cli` within itself instead of reusing existing CLI handlers. My thought is that the UI could instead pass parameters (like selected agents) into those praetorian-cli handlers, letting the interface handle selection while keeping execution logic centralized. This would avoid duplication and keep behavior consistent across both UI and CLI. SSH command is a great example as its duplicated in the normal cli and in the UI commands. Maybe using subprocess?

**Small items**

* Add `--details` option for capability descriptions, possibly with filters.
* Replace magic constants with centralized constants. (Theres a lot of 'health_check' for example)
* Make `Agent` a full type instead of a dictionary. (This will help solve the above)

DS-KoolAid committed:
refactoring and addressing comments

DS-KoolAid committed:
adding interface adapters to bridge cli and tui

DS-KoolAid commented:
@EvanLeleux I addressed your comments and refactored the TUI a bit.

As for the concern about duplication, I agree with you. But I don't love the subprocess approach. Using subprocess would sacrifice the TUI's ability to provide rich formatting, manage state (like selected agents), handle interactive features like progress bars and prompts, and would add performance overhead for every command. Instead, I moved the majority of the logic out of the TUI and into the SDK - as you suggested - and then created interface adapters that both the CLI and TUI rely on for command handling.

See image for high level overview (thanks claude): 

<img width="372" height="421" alt="image" src="https://github.com/user-attachments/assets/6ef85a3b-2ae1-4c24-91d0-ea1c04af8cf2" />

DS-KoolAid committed:
moving jwt decode out

EvanLeleux commented on praetorian_cli/handlers/aegis.py:
```diff
@@ -0,0 +1,117 @@
+    
+    # Get the shared command and invoke it
+    shared_cmd = aegis_shared.get_command(ctx, 'list')
+    ctx.invoke(shared_cmd, details=details, filter=filter)
```
This one is also defined in `list.py`, which should we keep?

EvanLeleux committed:
refactor with cc

EvanLeleux committed:
simplify

EvanLeleux committed:
🔖 Task: # ⚡️ SPARC Orchestrator - Task Execution  ## 🎯 Your Mission Build exactly what the user reque...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-142314

EvanLeleux committed:
🔖 Task: Why do we need to AegisCommandCore? I really just want to simplify this and make sure in the future ...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-150532

EvanLeleux committed:
🔖 Task: No, we should (for now) keep in line with the rest of the codebase. Like how add handler and list ha...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-150803

EvanLeleux committed:
🔖 Task: Yes I like this approach so much better. Could we even just have the Aegis Menu basically act as a "...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-151057

EvanLeleux committed:
🔖 Task: I love this plan. One last question, is there another way we can do handle_command in the TUI to not...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-151337

EvanLeleux committed:
🔖 Task: I love this. One more challenge, could instead of the command registry being a static dict, can we g...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-151614

EvanLeleux committed:
🔖 Task: Hmm, I dont like how complex invoke_click_command is now. Do we have to parse the params or can we j...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-152700

EvanLeleux committed:
🔖 Task: Lets do this, aegis_core should just be what the SDK is no need for another folder and type of scrip...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-153059

EvanLeleux committed:
🔖 Task: This is still too complex. I guess we need to find the right middle group between how it is now and ...

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-154128

EvanLeleux committed:
🔖 Task: Excellent lets get to implementing this straightforward approach ...

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-160541

EvanLeleux committed:
remove extra files

EvanLeleux committed:
remove some more

EvanLeleux committed:
remove even more

EvanLeleux committed:
🔖 Task: 1. When I run praetorian chariot aegis and the agents list that default runs fails, the display is n...

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/sdk/entities/aegis.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/menu.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🔖 Checkpoint: Edit /Users/evanleleux/Desktop/workspace/eng/praetorian-cli/praetorian_cli/ui/aegis/completion/set_completer.py

EvanLeleux committed:
🏁 Session end checkpoint: session-20250821-162722

EvanLeleux committed:
remove again

EvanLeleux committed:
remove more

EvanLeleux committed:
again

EvanLeleux committed:
🔖 Task: Just use git diff instead against the current branch evan/pr-edits ...

EvanLeleux committed:
simplify

EvanLeleux committed:
save work

EvanLeleux committed:
break out files

EvanLeleux committed:
reduce

EvanLeleux committed:
better ssh

EvanLeleux committed:
add tests

EvanLeleux committed:
warning ignore

DS-KoolAid committed:
Fix aegis.py functionality and enhance job command with interactive features

- Fixed list() method to return tuple (agents, None) for compatibility
- Improved error handling in get_capabilities and get_available_ad_domains
- Enhanced job run command with interactive capability picker
- Added OS detection and domain selection for better job targeting
- Added credential prompting for authentication-required capabilities
- Updated menu.py to use new tuple return format

DS-KoolAid committed:
Fix SSH user name extraction

- Auto-extract username using self.api.get_current_user() when none provided
- Always show user in SSH connection display info
- Improve authorized users checking to work without explicit user
- Ensure SSH command always includes username@hostname format

Matches functionality from mattj/aegis-integration branch

DS-KoolAid committed:
Enhance info command with mattj branch formatting

- Replace simple to_detailed_string() with comprehensive agent details
- Add system information display with OS, architecture, IPs, client ID
- Show network interfaces with IP addresses and interface names
- Display tunnel status with name, public hostname, authorized users
- Add --raw/-r flag support for JSON output
- Clear screen and show clean formatted output like mattj branch
- Update help command to document new info options and examples

DS-KoolAid committed:
Improve SSH argument parsing using shared handlers

- Replace manual SSH option parsing with SSHArgumentParser from ssh_utils
- More robust and consistent argument handling for SSH options
- Better error handling for malformed SSH arguments
- Maintains compatibility with existing CLI and UI callers

DS-KoolAid committed:
Update to professional Praetorian color scheme

- Replace basic colors with professional Praetorian purple theme from mattj branch
- Add new color variants: secondary, dark, dark_sec, text
- Improve visual consistency and professional appearance
- Better contrast and readability for UI elements

DS-KoolAid committed:
Eliminate SSH argument parsing duplication across layers

- Remove manual user parsing from UI ssh.py - use SSHArgumentParser instead
- Add agent validation in UI layer using shared utilities
- Simplify SDK parsing to only extract display info, not re-parse options
- Better error handling with console integration
- Consistent SSH option handling across CLI and UI interfaces

Fixes duplication between ui/ssh.py, handlers/aegis.py, and sdk/aegis.py

DS-KoolAid committed:
removing redundant import

DS-KoolAid committed:
Merge pull request #115 from praetorian-inc/evan/pr-edits

TUI Edits

EvanLeleux committed:
fix tests

EvanLeleux committed:
tidy
